docs/concurrency.org

```diff
@@ -1,68 +1,251 @@
+- preemptive multitasking ou cooperative (TODO JE NE SAIS TOUJOURS PAS psq le runtime peut toujours les preempter)
 
 *** Fibers
-- Same as green threads but cooperative multitasking instead of preemptive multitasking
+Same as green threads but cooperative multitasking instead of preemptive multitasking
 
 *** Preemptive multitasking
-- The underlying architecture (NOT us, but the OS for exemple) is going to be deciding what to execute and when
-- This means that our threads can be interrupted at any time
+In preemptive multitasking it is the underlying architecture (not us, but the OS or the runtime for exemple) that is in charge of choosing which threads to execute and when. This implies that our threads can be stopped (preempted) at any time, even if in the middle of a task. This method gives the advantage of not having to worr about a thread being starved, since the underlying architecture is going to make sure that everyone gets enough CPU time.
 
 *** Cooperative multitasking
-- A thread will continue to run until it decides to give up control (yielding)
-- We are in full control of the execution flow of our threads
-- The drawback of this method is that YOU have to think and code in a way to not have starving threads for exemple
-
-*** Thread
-- Multithreading
-- Each threads are still blocking
-- Main and historical way of doing concurrency
-- Parallelism (eventually): Threads allow for true parallel execution on multi-core CPUs, enabling better utilization of hardware resources and potentially improving performance.
-- Isolation: Threads provide strong isolation between concurrent tasks, with each thread having its own execution context, stack, and local variables.
-- Scalability: Managing a large number of threads can be challenging and may lead to scalability issues, as the operating system kernel has to allocate resources for each thread.
-- Usually used with a thread pool to avoid the overhead of creating and destroying threads for each task which is very expensive
-
-*** Event-driven programming (ex: NodeJS)
-- Event loop that regularly checks if "events" are launched
-- Scalability: Event-driven architectures are inherently scalable, as they can handle a large number of concurrent connections with low memory and CPU overhead.
-- Code behind NodeJS: https://github.com/libuv/libuv
-
-** Zig current state
+Contrary to preemptive multitasking, it is the progammer job to choose which and when the differents threads are executed. Threads are going to run until they are explicitly yielding control back. This method gives the advantage to have the progammer to have a fine grained control over his ressources, but also implies that the programmer has to think about not starving threads.
+
+*** Thread (ChatGPT)
```

Comment:
Why there is ChatGPT ?

Reply:
because i initially wrote it with chatgpt and put that to remember that I have to change a bit, changed it btw

Reply:
Ok it's cool that we're keeping the fact that we say which section uses chatgpt

Reply:
I completly rewrote it without chatgpt, its cleaner like that :)